rust/src/librarydb.rs

```diff
@@ -0,0 +1,371 @@
+//! This module implements a thread safe library database.
+//!
+//! A library database contains a collection of libraries.
+//! A library contains a collection of file entries.
+//! Libraries that contain multiple ok entries with the same case-insensitive file path are rejected.
+//!
+//! The order of the libraries matters.
+//! When a file path matches multiple entries, the first entry shadows the other entries.
+//!
+//!
+//! # Rustified
+//!
+//! This module is a rustified version of `src/sgp/LibraryDataBase.cc` and `src/sgp/LibraryDataBase.h`.
+//!
+//! The main differences are:
+//!  * it is thread safe
+//!  * shadowing respects library order and uses the full file path,
+//!    the original shadowed libraries based on the longest base path
+//!
+//!
+//! # FFI
+//!
+//! [`stracciatella::c::librarydb`] contains a C interface for this module.
+//!
+//! [`stracciatella::c::librarydb`]: ../c/librarydb/index.html
+
+use std::cmp;
+use std::collections::HashSet;
+use std::fmt;
+use std::fs::File;
+use std::io::{self, Seek, SeekFrom};
+use std::ops;
+use std::path::{Component, Path, PathBuf};
+use std::sync::{Arc, Mutex, MutexGuard, RwLock};
+
+use crate::file_formats::slf::{SlfEntryState, SlfHeader};
+
+/// Thread safe library database.
+#[derive(Debug)]
+pub struct LibraryDB {
+    inner: Mutex<LibraryDBInner>,
+}
+
+/// Library database.
+#[derive(Debug)]
+pub struct LibraryDBInner {
+    /// Thread safe libraries.
+    arc_libraries: Vec<Arc<RwLock<Library>>>,
+}
+
+/// Library.
+/// Keeps the library file open for reading and a list of entries that are ok.
+/// The library file contains the data of each entry.
+#[derive(Debug)]
+pub struct Library {
+    /// Real path of the library file for display purposes.
+    library_path: PathBuf,
+    /// Library file open for reading.
+    library_file: File,
+    /// Base path of the entries in the library.
+    base_path: String,
+    /// List of ok entries in the library.
+    entries: Vec<LibraryEntry>,
+}
+
+/// Library entry.
+/// Represents a file in a library.
+#[derive(Debug)]
+pub struct LibraryEntry {
+    /// File path without the base path of the library.
+    file_path: String,
+    /// Start of the file data in the library file.
+    data_start: u64,
+    /// End of the file data in the library file.
+    data_end: u64,
+}
+
+/// Library file.
+/// Provides access to the data of a library entry through Read and Seek.
+#[derive(Debug)]
+pub struct LibraryFile {
+    /// The index of the entry of this file.
+    index: usize,
+    /// The thread safe library of this file.
+    arc_library: Arc<RwLock<Library>>,
```

Comment:
It would also need a mutex to ensure thread safety from C calls.
Not being linked to the library means not being able to implement a writable library file (not implemented).
Is it desirable to make it self contained? I think it's better like this, I can even check that there are no open library files when the library is dropped (artificial restriction).

Reply:
And it would also need the file handle cloned or reopened for itself.

Reply:
Fine for me to keep it that way...